sphinx_immaterial/content_tabs.py

```diff
@@ -0,0 +1,220 @@
+        attributes["checked"] = "checked"
+    self.body.append(self.starttag(node, "input", **attributes))
+
+
+def depart_tab_input(self, node):
+    self.body.append("</input>")
+
+
+def visit_tab_label(self, node):
+    attributes = {"for": node["input_id"]}
+    self.body.append(self.starttag(node, "label", **attributes))
+
+
+def depart_tab_label(self, node):
+    self.body.append("</label>")
+
+
+def visit_tab_set(self, node):
+    attributes = {"data-tabs": node["data-tabs"]}
+    self.body.append(self.starttag(node, "div", **attributes))
+
+
+def depart_tab_set(self, node):
+    self.body.append("</div>")
+
+
+class MaterialTabSetHtmlTransform(SphinxPostTransform):
+    """Transform tab-set to HTML specific AST structure."""
+
+    default_priority = 200
+    formats = ("html",)
+
+    def run(self, **kwargs: Any) -> None:
+        """Run the transform."""
+        matcher = NodeMatcher(nodes.container, type="md-tab-set")
+        for set_numb, tab_set in enumerate(self.document.traverse(matcher)):
+            tab_set_identity = f"__tabbed_{set_numb + 1}"
+            children = []
+            tab_label_div = nodes.container("", is_div=True, classes=["tabbed-labels"])
+            tab_content_div = nodes.container(
+                "", is_div=True, classes=["tabbed-content"]
+            )
+            tab_total = 0
+            for tab_item in tab_set.children:
+                try:
+                    tab_label, tab_block = tab_item.children
+                except ValueError:
+                    print(tab_item)
+                    raise
+                tab_total += 1
+                tab_item_identity = tab_set_identity + f"_{tab_total}"
+
+                # create: <input checked="checked" id="id" type="radio">
+                input_node = content_tab_input(
+                    "",
+                    id=tab_item_identity,
+                    set_id=tab_set_identity,
+                    type="radio",
+                    checked=(tab_total == 1),
+                )
+                input_node.source, input_node.line = tab_item.source, tab_item.line
+                children.append(input_node)
+
+                # create: <label for="id">...</label>
+                label_node = content_tab_label(
+                    "",
+                    *tab_label.children,
+                    input_id=tab_item_identity,
+                    classes=tab_label["classes"],
+                )
+                label_node.source, label_node.line = tab_item.source, tab_item.line
+                tab_label_div += label_node
+
+                # add content
+                tab_content_div += tab_block
+
+            tab_set.children = children
+            tab_set += tab_label_div
+            tab_set += tab_content_div
+
+
+def setup(app: Sphinx) -> None:
+    app.add_directive("md-tab-set", MaterialTabSetDirective)
+    app.add_directive("md-tab-item", MaterialTabItemDirective)
+    app.add_post_transform(MaterialTabSetHtmlTransform)
```

Comment:
I think it would be best to avoid using `add_post_transform`, since that means an additional full traversal of every document.  I think the main thing you are using the post transform for is assigning a unique index to each tab set --- for that you could instead use `env.temp_data` (which is cleared for each document) or handle it in the writer (i.e. the `visit_` functions), where you could store the counter as an attribute on the writer.

Reply:
>  I think the main thing you are using the post transform for is assigning a unique index to each tab set

Not true. The transformer, takes the tab item and splits it up into a label, input, and content's div elements. For each tab set the labels and and content's div elements go in their own div containers, while the input elements go directly after the div for the tab set but before the label and content groups.

I couldn't think of a way to organize these elements from the `MaterialTabItemDirective.run()` And, there was the additional pain of figuring out the `id`, `for`, and `name` attributes value; I didn't know about the `env.temp_data`.

Reply:
Okay, I see.  Thinking about this more, I think this could all be handled in `visit_tab_set`:

- Store the counter for ids in the HTMLTranslator (i.e. `self`)
- In `visit_tab_set`, emit the opening div, then emit the input tag for each child, then emit the opening tag for the labels.  Then emit the opening label tag for each child, and call `child_label.walkabout(self)` to render the label.
- After emitting the labels, do a similar thing for the content.
- Emit the closing tag.
- raise docutils.nodes.SkipNode to prevent docutils from itself recursing on the children.

Reply:
Interesting challenge. Much of this code is heavily modified from the `sphinx-design` ext. I didn't bother with attribution because the `sphinx-design` uses MIT lisence, and the code changed so much from the initial copy.

Clearly, I have a lot to learn about `docutils` - I wish they had better/complete docs.

Reply:
Yeah unfortunately the docs are rather limited --- I have found the only way to really understand docutils and sphinx is to study the source code.

Reply:
Agreed, its just **_a lot_** to study. Also, its not a good sign when a docs-based library lacks on its own docs... 🤣

Reply:
`temp_data` persists across all pages, so storing the tab set count in `temp_data` yields an `id` of `__tabbed_4` on the second set of tabs in content_tabs.rst. This is not consistent with upstream HTML generation. It doesn't seem to be a problem, but it is a less precise re-implementation. My main concern here is having to retool it when linked tabs moves to public releases in upstream.

### un-committed mod
Currently, I'm not using the `walkabout()` idea. I basically moved all the algorithm from the transformer into `visit_tab_set()`. Is there an objection to having separate `visit_*()` functions? I'm uneasy about writing all the raw HTML in a single `visit_*()` function.

Reply:
temp_data should be cleared after reading each document, but note that it only applies to the initial reading stage, i.e. when the directive is first being processed.  During later stages I don't think it gets reset, so it shouldn't be used for per-document state.  During writing, you can instead use the HTMLTranslator object itself (i.e. the `self` parameter of the visit methods) to store per-document state, because a separate HTMLTranslator object is created per document.

Reply:
I can't treat the translator like a dict; that results in a error. I'm currently doing
```py

def visit_tab_set(self: HTMLTranslator, node: content_tab_set):
    # increment tab set counter
    temp_data = self.builder.env.temp_data
    temp_data["tab_set_count"] = temp_data.get("tab_set_count", 0) + 1
```
Even dynamic attribute creation has the same effect
```python
def visit_tab_set(self: HTMLTranslator, node: content_tab_set):
    # increment tab set counter
    self.tab_set_count = getattr(self, "tab_set_count", 0) + 1
```
The old way i was using was a global 1-item list in content_tabs.py that did get reset for each page.

Reply:
NM, dynamic attribute creation worked I was just misreading the HTML... commit incoming.